odl/tomo/geometry/geometry.py

```diff
@@ -0,0 +1,383 @@
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Geometry base class."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from abc import ABCMeta, abstractmethod
+from future import standard_library
+standard_library.install_aliases()
+from builtins import object, super
+
+# External
+import numpy as np
+
+# Internal
+from odl.util.utility import with_metaclass
+from odl.discr.grid import TensorGrid
+from odl.tomo.geometry.detector import Detector
+
+__all__ = ('Geometry', 'DivergentBeamGeometry', 'AxisOrientedGeometry')
+
+
+class Geometry(with_metaclass(ABCMeta, object)):
+
+    """Abstract geometry class.
+
+    A geometry is described by
+
+    * a detector,
+    * a set of detector motion parameters,
+    * a function mapping motion parameters to the location of a
+      reference point (e.g. the center of the detector surface),
+    * a rotation applied to the detector surface, depending on the motion
+      parameters,
+    * a mapping from the motion and surface parameters to the detector pixel
+      direction to the source,
+    * optionally a mapping from the motion parameters to the source position
+    """
+
+    def __init__(self, ndim, motion_grid, detector):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        ndim : `int`
+           The number of dimensions of the geometry
+        motion_grid : `TensorGrid`
+           Grid for the motion of the detector
+        detector : `Detector`
+           The detector
+        """
+        if ndim < 1:
+            raise ValueError('ndim {!r} not a positive integer.'
+                             ''.format(ndim))
+
+        if not isinstance(motion_grid, TensorGrid):
+            raise TypeError('motion_grid {!r} not a `TensorGrid` instance.'
+                            ''.format(motion_grid))
+
+        if not isinstance(detector, Detector):
+            raise ValueError('motion_grid {!r} not a `Detector` instance.'
+                             ''.format(detector))
+
+        self._ndim = int(ndim)
+        self._motion_grid = motion_grid
+        self._motion_params = motion_grid.convex_hull()
+        self._detector = detector
+        self._implementation_cache = {}
+
+    @property
+    def motion_params(self):
+        """The motion parameters given as an `IntervalProd`."""
```

Comment:
These accessors are in general quite important and could use somewhat longer docstrings. What is a motion parameter? Perhaps link to some doc (we could make an issue on this)